testing/trino-testing/src/main/java/io/trino/testing/containers/TestContainer.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testing.containers;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.net.HostAndPort;
+import io.airlift.log.Logger;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.startupcheck.IsRunningStartupCheckStrategy;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.MountableFile;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.net.HostAndPort.fromParts;
+import static java.util.Objects.requireNonNull;
+
+public abstract class TestContainer
+        implements Closeable
+{
+    private final Logger log;
+
+    private final String hostName;
+    private final Set<Integer> ports;
+    private final int startupRetryLimit;
+    private final Map<String, String> filesToMount;
+    private final Map<String, String> envVars;
+    private final Optional<Network> network;
+
+    protected GenericContainer<?> container;
+
+    protected TestContainer(
+            String image,
+            String hostName,
+            Set<Integer> ports,
+            Map<String, String> filesToMount,
+            Map<String, String> envVars,
+            Optional<Network> network,
+            int startupRetryLimit)
+    {
+        this.log = Logger.get(this.getClass());
+        this.container = new GenericContainer<>(requireNonNull(image, "image is null"));
+        this.ports = requireNonNull(ports, "ports is null");
+        this.hostName = requireNonNull(hostName, "hostName is null");
+        checkArgument(startupRetryLimit > 0, "startupRetryLimit needs to be greater or equal to 0");
+        this.startupRetryLimit = startupRetryLimit;
+        this.filesToMount = requireNonNull(filesToMount, "filesToMount is null");
+        this.envVars = requireNonNull(envVars, "envVars is null");
+        this.network = requireNonNull(network, "network is null");
+        setupContainer();
+        startContainer();
```

Comment:
I would prefer to have caller explicitly call out to `start` (move it out of object constructor).